scala/src/com/google/idea/blaze/scala/run/producers/ScalaTestContextProvider.java

```diff
@@ -76,32 +92,49 @@ private static ScClass getTestClass(ConfigurationContext context) {
+        .map(classWithTestName -> new ClassAndTestName(classWithTestName.testClass(), classWithTestName.testName().getOrElse(() -> null)))
+        .getOrElse(() -> null);
   }
 
-  @Nullable
-  private static ScClass getTestClass(Location<?> location) {
-    PsiElement element = location.getPsiElement();
-    ScClass testClass;
-    if (element instanceof ScClass) {
-      testClass = (ScClass) element;
-    } else {
-      testClass = PsiTreeUtil.getParentOfType(element, ScClass.class);
-    }
-    if (testClass != null && isTestClass(testClass)) {
-      return testClass;
-    }
-    return null;
-  }
+  static TestContext.BlazeFlagsModification setScalatestTestSelectorArguments(ClassAndTestName classAndTestName) {
+    String testClassSelectorFlag = BlazeFlags.TEST_ARG + "-s";
+    String testClassFlag = BlazeFlags.TEST_ARG + classAndTestName.testClass.getQualifiedName();
+    String testNameSelectorFlag = BlazeFlags.TEST_ARG + "-t";
+    // Scalatest names can contain spaces, so the name needs to be quoted
+    String testNameFlag = BlazeFlags.TEST_ARG + "\"" + classAndTestName.testName + "\"";
+    return new TestContext.BlazeFlagsModification() {
+      private boolean hasTestName() {
+        return classAndTestName.testName != null;
+      }
 
-  private static boolean isTestClass(ScClass testClass) {
-    TestFramework framework = TestFrameworks.detectFramework(testClass);
-    return framework instanceof ScalaTestTestFramework && framework.isTestClass(testClass);
-  }
+      @Override
+      public void modifyFlags(List<String> flags) {
+        // Ordering matters here. ScalaTest interprets "-s className -t testName" to mean "run the test in the selected class with the selected name".
+        // The reverse ordering "-t testName -s className" means "run all tests with the given name, and also all tests in the given class".
+        if (!flags.contains(testClassSelectorFlag) &&
+            !flags.contains(testClassFlag)) {
+          flags.add(testClassSelectorFlag);
+          flags.add(testClassFlag);
+        }
+
+        if (hasTestName()) {
+          if (!flags.contains(testNameSelectorFlag) &&
+              !flags.contains(testNameFlag)) {
+            flags.add(testNameSelectorFlag);
+            flags.add(testNameFlag);
+          }
+        }
+      }
 
-  private static String getTestFilter(PsiClass testClass) {
-    // TODO: may need to append '#' if implementation changes.
-    // https://github.com/bazelbuild/rules_scala/pull/216
-    return testClass.getQualifiedName();
+      @Override
+      public boolean matchesConfigState(RunConfigurationFlagsState state) {
+        List<String> rawFlags = state.getRawFlags();
+        boolean testClassFlagsSet = rawFlags.contains(testClassSelectorFlag) && rawFlags.contains(testClassFlag);
+        boolean relevantNameFlagsSet = !hasTestName() || (rawFlags.contains(testNameSelectorFlag) && rawFlags.contains(testNameFlag));
```

Comment:
Personally I'd prefer to check these flags more strictly. So to check whether -s is directly followed by suite name and -t by test case name. Theoretically a test could have the same name as a suite and this code would fail. It does not happen now only because we surround test name with double quotes.

Maybe we could extract and unit test this logic?

Reply:
Oh, and it's also wrong when you first try to run only Test1, and the whole Suite afterwards
```
class Suite extends AnyFunSuite {
  test("Test1") {

  }
  test("Test2") {

  }
}
```

Reply:
Did this. Let me know if you have a better place in mind for the flag modifier class, I wasn't sure where to put it so for now it's just a static inner class.

Reply:
> Oh, and it's also wrong when you first try to run only Test1, and the whole Suite afterwards

Good catch, it wasn't clear to me why matchesConfigState even needed to exist, but it makes sense if it's caching old flags.

So presumably in `modifyFlags`, I'd need to remove any previous occurrences of the `-s` and `-t` flags?

Reply:
I don't think so. If the flags are already there, something must have gone really bad, so I'd consider it a non-recoverable error and just throw, but I'm not sure what would happen then. Other solution is to indeed replace the flags, but log error with IntelliJ logging mechanism.

Reply:
I'd rather just leave it as is if the intended way these methods are called is that `modifyFlags` isn't supposed to get called on the same flag list twice.

Reply:
I think your example works now, at least it seems to when I run it. Running test 1 and then trying to run the test class seems like it works as expected for me.

Reply:
Resolving since the given example works now